namenode/namenode.go

```diff
@@ -44,27 +55,279 @@ func (server *NameNode) Create(path string, success *bool) error {
 
 func (server *NameNode) Exists(path string, exists *bool) error {
 	slog.Info("Exists request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	_, *exists = server.inodes[path]
 	return nil
 }
 
+// TODO: 2pc, notify all the datanodes to delete the blocks before responding
 // This RPC doesn't return anything
 func (server *NameNode) Delete(path string, unused *bool) error {
 	slog.Info("Delete request", "path", path)
 
-	server.globalRWLock.Lock()
-	defer server.globalRWLock.Unlock()
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
 
 	delete(server.inodes, path)
 	return nil
 }
 
 func (server *NameNode) GetAttributes(path string, fileAttributes *common.FileAttributes) error {
 	slog.Info("GetAttributes request", "path", path)
-	server.globalRWLock.RLock()
-	defer server.globalRWLock.RUnlock()
+	server.inodeRWLock.RLock()
+	defer server.inodeRWLock.RUnlock()
 	*fileAttributes = server.inodes[path].fileAttributes
 	return nil
 }
+
+func (server *NameNode) RegisterDataNode(dataNodeEndpoint string, success *bool) error {
+	slog.Info("RegisterDataNode request", "dataNodeEndpoint", dataNodeEndpoint)
+
+	server.inodeRWLock.Lock()
+	defer server.inodeRWLock.Unlock()
+
+	// TODO: Check if the datanode is already registered
+	server.datanodes[dataNodeEndpoint] = DataNodeInfo{
+		heartbeatReceived: true,
+		isAlive:           true,
+	}
+
+	*success = true
+	return nil
+}
+
+// NameNode receives block report from DataNode,
+// and updates all block storage info in the corresponding inode
+func (server *NameNode) ReportBlock(blockReport common.BlockReport, success *bool) error {
+	slog.Info("ReportBlock request", "dataNodeEndpoint", blockReport.Endpoint,
+		slog.Any("Block MetaData", blockReport.BlockMetadata))
+
+	server.inodeRWLock.Lock()
```

Comment:
This is something we should try to avoid.

Reply:
This basically stops the world

Reply:
Although block report doesn't occur frequently, this operation is too disruptive.

Reply:
I tried to avoid it but I think it is hard to avoid

Reply:
Let me try to find a better way. Leave it now.

Reply:
by the way, if we change some variable inside one inode, do we need to use `server.inodeRWLock.Lock()`? Or is it only needed when we add or delete a file? If it is the latter case, then the lock can be applied only to the lines in "// If the file doesn't exist, create a new inode" part

Reply:
Also I think it is reasonable for BlockReport to be costly. It is more important than create / delete in our design, it doesn't occur frequently. And for example, when a namenode reboots, it needs n-2 BlockReport before receiving request from clients, so it won't hurt performance too much in this case.